design/open/repository-operations-error-handling.md

```diff
@@ -0,0 +1,121 @@
+  * If failed return the error
+* Generate the partition key for the `Repository`, from its `RepositoryID` and `unique_identifier`
+* Scan through the branches in the partition partition and for each `Branch`:
+  * Delete all the relevant staged objects (based on `staging_token` and `sealed_tokens`) - same as with `DeleteBranch`
+  * If an error occurs at each step, return the error - the `Repository` is already marked as deleted and so it and all its correlated entities are unreachable
+* Delete the `Branch`
+  * If failed - return the error
+* Scan through the rest of the partition and delete all objects. That is `Commit`s and `Tag`s
+  * With any failure, return the error
+* If all deletions above are successful, delete the `Repository` itself
+  * If failed return the error
+
+### ListRepositories
+* List repositories should only return `active` entities. That can be achieved by identifying and skipping `Repository` entries with status other than `active`
+
+### Other Operations During Repository Creation/Deletion
+* All operations that require a valid `Repository` should start with `GetRepository`. This will prevent the usage of `Repository` in mid creation (with SQL DB it is impossible to reach this entity)
+* If a `Repository` is being deleted when it is assumed to exists (e.g. `CreateBranch` operation that starts when the `Repository` exists and `active`, then the `Repository` is removed and only then the `Branch` is created in the KV) will leave the created entity unreachable, as the `Repository` does no longer exist (or is not `active` anymore) and so, though the newly created entity exists, it exists as unreachable garbage. This is similar of an entity being created, than the `Repository` is deleted and the entity removed with it
+
+## Garbage Collection - Nice to Have
+The above proposed solution provides the correctness needed for graveler. In addition, a dedicated garbage collection can help and maintain the KV Store cleaner and, as a result, smaller.
+The logic is very simple, given the solution above is implemented:
+* Scan through all the `Repository` entities in the common `graveler` partition
+* For each `Repository`:
+  * If it is in `active` state - ignore
+  * If it is in `deleted` or `failed` state - delete it applying the same steps as described above for `DeleteRepository`
+  * If it is in `initial` state - decide if it should move to `failed` and if so, set to `failed` and try to delete, as described above
+Run periodically (**TBD** How often?) as a background worker and upon error just skip the `Repository`.
+
+## Open Questions
+1. What error value should we return for a `GetRepository` when the repository is in `deleted` or `failed` state?
+  * `ErrRepositoryNotFound`? (might be a problem as a user may attempt to create such a `Repository` and fail).
+  * Some new `ErrResourceBusy`/`ErrRepositoryDeletionInProgress`?
+2. `Repository`'s `unique_identifier` - nanoid? xid? `RepositoryID`+creation-timestamp?
```

Comment:
I think this is an implementation details that is out of the scope of the design